ts_src/psetv2/finalizer.ts

```diff
@@ -0,0 +1,249 @@
+import { payments } from '..';
+import { witnessStackToScriptWitness } from '../psbt';
+import { Input } from './input';
+import { PartialSig } from './interfaces';
+import { Pset } from './pset';
+import { classifyScript, getPayment, hasSigs, isP2WPKH } from './utils';
+
+export type FinalScriptsFunc = (
+  inputIndex: number, // Which input is it?
+  input: Input, // The PSBT input contents
+  script: Buffer, // The "meaningful" locking script Buffer (redeemScript for P2SH etc.)
+  isSegwit: boolean, // Is it segwit?
```

Comment:
In bitcoinjs-lib, they choose to define another type `FinalTaprootScriptFunc` instead of re-using `FinalScriptFunc` in order to add brand new code for taproot support, instead of updating the existing one. Much easier to review and maintain for a quite big codebase.

My initial idea was to follow their lead. Still, since this is a completely new sub-package and it will most likely be different from any future psbtv2, if you want to make a proposal on how to abstract the inputs finalization, I'd be happy to discuss it.

PS. I've already started to add support for taproot on psetv2, and didn't change the finalizer yet, therefore this is the best time for a new proposal.

Reply:
I don't really like the idea of two different functions types. What if:
```typescript
type FinalizeFunction = (inputIndex: number, pset: Pset) => {  finalScriptSig?: Buffer; finalScriptWitness?: Buffer; }
```

_why pset as parameter?_
* it includes all the input data (taproot fields, witnessScript, redeemScript etc...)
* elements tapscript with introspection finalizers may need extra data from the transaction (outputs, timelock etc...)

_why no script type?_
* We could include the "meaningful" `script` and `is*` parameters but IMO the "script type" check could be done outside of the finalize function if we write 1 FinalizeFunctionFactory for each type (and a global one, checking the script type and calling the associated finalizer?)

something like:
```typescript
const finalizerP2SH = (redeemScript: Buffer): FinalizeFunction => {}
const finalizerP2WSH = (redemScript:  Buffer, witnessScript: Buffer): FinalizeFunction => {}

const defaultFinalizer: FinalizeFunction = (inputIndex: number, pset: Pset) => {
   const input = pset.ins[inputIndex] // get the input to sign
   switch (getTypeOfScriptToSign(input)) {
      case P2SH:
           return finalizerP2SH(input.redeemScript)
      // etc...
   }
} 
```